src/main/java/com/google/devtools/build/lib/skyframe/BzlLoadFunction.java

```diff
@@ -944,12 +947,12 @@ private static RepositoryMapping getRepositoryMapping(
     }
 
     if (key instanceof BzlLoadValue.KeyForBzlmod) {
-      if (repoName.equals(RepositoryName.BAZEL_TOOLS)) {
-        // Special case: we're only here to get the @bazel_tools repo (for example, for
-        // http_archive). This repo shouldn't have visibility into anything else (during repo
-        // generation), so we just return an empty repo mapping.
-        // TODO(wyv): disallow fallback.
-        return RepositoryMapping.ALWAYS_FALLBACK;
+      if (key instanceof BzlLoadValue.KeyForBzlmodBootstrap) {
```

Comment:
instead of adding a new key type, can we just check `isFileSafeForUninjectedEvaluation` here? the method name might need updating, but it's the same special case AFAIU.

Reply:
Would this work with `http_jar` though? It is defined in the same file as `http_archive`, but needs to resolve a `Label` in `rules_java`: https://github.com/bazelbuild/bazel/pull/19997/files#diff-94940b11687a256a73671f698151c5ad9a71f820ca9295efbd8c0506a6433da9

Reply:
I can look into adding test coverage for this case in either of these PRs.

Reply:
ah, you're right, no it won't. But that one looks very weird to me already -- it's awkward that somehow using a repo rule requires having a dependency on a certain other module... I guess we'll have to move `http_jar` out of bazel_tools some day if we want to move rules_java out.

Reply:
for the short term we could also move the `_http_jar_impl` function into another package (`tools/build_defs/repo/private` or some such). WDYT?

Reply:
> But that one looks very weird to me already -- it's awkward that somehow using a repo rule requires having a dependency on a certain other module... I guess we'll have to move http_jar out of bazel_tools some day if we want to move rules_java out.

I need think a bit more about how this could be cleanly resolved (or if it's possible at all...)

Reply:
Ah, I see, we could use a separate `RepositoryMapping` for just this single file and not necessarily its entire transitive closure, right? If that's true, then this approach is definitely better and I can make that change.

Reply:
yeah, `getRepositoryMapping` is scoped to a single bzl load, so it does work if you put stuff that needs a real repo mapping into a subpackage.